src/graph/LookupExecutor.cpp

```diff
@@ -0,0 +1,555 @@
+            auto* left = rExpr->left();
+            auto* right = rExpr->right();
+            /**
+             *  TODO (sky) : Support WHERE ：tag1.col2 != tag1.col3
+             *  Handler error for FuncExpr or  ArithmeticExpr contains
+             *  AliasPropExpr , for example :
+             *  WHERE lookup_tag_2.col2 > (lookup_tag_2.col3 - 100)
+             */
+            if (left->kind() == nebula::Expression::kAliasProp &&
+                right->kind() == nebula::Expression::kAliasProp) {
+                return Status::SyntaxError("Syntax error ：%s", rExpr->toString().c_str());
+            }  else if (left->kind() == nebula::Expression::kAliasProp) {
+                auto* aExpr = dynamic_cast<const AliasPropertyExpression*>(left);
+                prop = *aExpr->prop();
+                filters_.emplace_back(std::make_pair(prop, rExpr->op()));
+            } else if (right->kind() == nebula::Expression::kAliasProp) {
+                auto* aExpr = dynamic_cast<const AliasPropertyExpression*>(right);
+                prop = *aExpr->prop();
+                filters_.emplace_back(std::make_pair(prop, rExpr->op()));
+            } else {
+                skipOptimize_ = true;
+            }
+            break;
+        }
+        case nebula::Expression::kFunctionCall : {
+            auto* fExpr = dynamic_cast<const FunctionCallExpression*>(expr);
+            auto* name = fExpr->name();
+            if (*name == "udf_is_in") {
+                return Status::SyntaxError("Unsupported function ： %s", name->c_str());
+            }
+            break;
+        }
+        default : {
+            return Status::SyntaxError("Syntax error ： %s", expr->toString().c_str());
+        }
+    }
+    return Status::OK();
+}
+
+Status LookupExecutor::checkFilter() {
+    auto status = traversalExpr(sentence_->whereClause()->filter());
+    return status;
+}
+
+Status LookupExecutor::chooseIndex() {
+    auto ret = (isEdge_) ? ectx()->getMetaClient()->getEdgeIndexesFromCache(spaceId_)
+                         : ectx()->getMetaClient()->getTagIndexesFromCache(spaceId_);
+    if (!ret.ok()) {
+        LOG(ERROR) << "No index was found";
+        return Status::IndexNotFound();
+    }
+
+    auto indexes = ret.value();
+    if (indexes.empty()) {
+        return Status::IndexNotFound();
+    }
+
+    auto validIndexes = findValidIndex(indexes);
+    if (validIndexes.empty()) {
+        LOG(ERROR) << "No matching index was found";
+        return Status::Error("No matching index was found");
+    }
+    return findOptimalIndex(validIndexes);
+}
+
+/**
+ * Choose the index that filter best matches, for example :
+ * the filter is col1 == 1 and col2 == 2 and col3 == 3,
+ * the indexes are index1 (col4, col3, col2, col1)
+ *                 index2 (col1, col2, col3, col4)
+ *                 index3 (col1, colA, col2, col3)
+ * as above, these indexes all are valid, but the index2 should be choose.
```

Comment:
In your case,  for the expression " col1 == 1 and col2 == 2 and col3 == 3",   only index2 is valid. 
index1 and index3 can't be used.